### PR TITLE
Feature - Support CSS styles

### PR DESCRIPTION
**Pull request by lowlighter**
This implement what was suggested in https://github.com/markmead/alpinejs-component/issues/8#issuecomment-1699941700

I tested it and it works quite well, here's a small demo if you want to preview it:
https://jsfiddle.net/La3nvfey/

I think the only caveat is that the shadow root is not aware of its parents, which means that rules based on hierarchy won't work.
I'm not sure what we could do to solve this (maybe some workaround where we use [`Element.matches`](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches) and we iterate through ancestors to find matching rules ?)

Since it may not be wanted by everyone and I'm not sure of performances on large sets, I gated it behind a `styles` params

### CODE REVIEW

**Comment by markmead** [comment]
Interesting, that's the first time I've seen that comment you linked. Shame I wasn't @'d otherwise I'd have seen it earlier 🤷‍♀️ 

~It would be great to see a PR if you have time?~ Ignore me, ha! 😅

File: src/index.js [hunk 1]
```diff
@@ -5,12 +5,26 @@ export default function (Alpine) {
 
       const hasDynamicTemplate = this.hasAttribute(':template')
       const hasDynamicUrl = this.hasAttribute(':url')
+      const hasDynamicStyles = this.hasAttribute(':styles')
```
Comment (markmead) [inline src/index.js]: I don't think we need to support `:styles` so this can be removed.

File: src/index.js [hunk 2]
```diff
@@ -5,12 +5,26 @@ export default function (Alpine) {
...
+        styles.insertRule(':host { display: contents; }')
+        let i = 0
+        for (const rule of [...document.styleSheets].flatMap((sheet) => [...sheet.cssRules])) {
+          if (rule instanceof CSSStyleRule && rule.selectorText === ':root') {
+            continue
+          }
+          styles.insertRule(rule.cssText, i++)
+        }
+        shadowDom.adoptedStyleSheets = [styles]
+      }
```
Comment (markmead) [inline src/index.js]: This looks a little messy, I think Prettier will fix that but do we need the `i++`? It feels like computer speak.

File: README.md [hunk 3]
```diff
@@ -173,6 +173,11 @@ component itself.
 </div>
 ```
 
+You can also use the `styles` attributes to make the "Shadow DOM" inherits the global CSS:
```
Comment (markmead) [inline README.md]: Do you think a comment is needed in the inline styles section of the README about inline styles, or do you think it's safe to remove that section altogether?

File: src/index.js (hunk 2 above)
Comment (markmead) [reply src/index.js]: Do we even need `i++`? I just rewrote it like this and it seems to be working OK:

```js
if (componentStyles) {
  const newStyle = new CSSStyleSheet()
  const documentStyles = [...document.styleSheets].flatMap(({ cssRules }) => [...cssRules])

  for (const styleRule of documentStyles) {
    if (styleRule instanceof CSSStyleRule && styleRule.selectorText === ':root') {
      continue
    }

    newStyle.insertRule(styleRule.cssText)
  }

  shadowDom.adoptedStyleSheets = [newStyle]
}
```

**Comment by markmead** [comment]
Pushed a commit with some changes, if you could take a look and see what you think. Feel free to make changes.

File: src/index.js (hunk 1 above)
Comment (lowlighter) [reply src/index.js]: I guess we could enable it by default, but I'd consider this a breaking change if users were already using the workaround. Also maybe it's be worth to investigate if we can emulate the hierarchy rules if the aim is to make a truly *transparent* support 🤔 ?

File: src/index.js (hunk 2 above)
Comment (lowlighter) [reply src/index.js]: Yeah I'll run prettier on it to match the rest of the formatting 👍 

As for the index, it is now optional [`CSSStyleSheet.insertRule#parameters`](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/insertRule#parameters) but in the past it was a mandatory param (here's the browser [compatibility list](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/insertRule#browser_compatibility), though I don't think it'd impact a lot of people, the browsers versions are all pretty old

Comment (markmead) [reply src/index.js]: If you pull down the latest it should have all the prettier updates.

File: src/index.js (hunk 1 above)
Comment (markmead) [reply src/index.js]: Attributes with `:` at the start are for values that will be passed as a variable. It's not something I see people doing and has been removed in the commit I did the other day. 

As for the hierarchy rules I think leave it for now and see if people request it.

**Comment by lowlighter** [comment]
Nice it looks much better indeed !
Thanks a lot for adding this feature ❤️ !

**Pull request merged**